lib/benchmark/bench.go

```diff
@@ -74,160 +76,140 @@ func (c *Config) Benchmark(ctx context.Context, tc *client.TeleportClient) (Resu
-	result.Histogram = hdrhistogram.New(MinValue, MaxValue, SignificantFigures)
-	results := make([]benchMeasure, 0, c.MinimumMeasurements)
+	result.ResponseHistogram = hdrhistogram.New(MinValue, MaxValue, SignificantFigures)
+	result.ServiceHistogram = hdrhistogram.New(MinValue, MaxValue, SignificantFigures)
+	results := make([]*benchMeasure, 0, c.MinimumMeasurements)
 	statusTicker := time.NewTicker(1 * time.Second)
 	timeElapsed := false
 	start := time.Now()
-
 	for {
 		if c.MinimumWindow <= time.Since(start) {
 			timeElapsed = true
 		}
 		select {
-		case measure := <-responseC:
-			result.Histogram.RecordValue(int64(measure.End.Sub(measure.Start) / time.Millisecond))
+		case measure := <-resultC:
 			results = append(results, measure)
 			if timeElapsed && len(results) >= c.MinimumMeasurements {
-				go cancelWorkers()
+				go cancel()
 			}
 			if measure.Error != nil {
 				result.RequestsFailed++
 				result.LastError = measure.Error
 			}
+			result.ResponseHistogram.RecordValue(int64(measure.End.Sub(measure.ResponseStart) / time.Millisecond))
+			result.ServiceHistogram.RecordValue(int64(measure.End.Sub(measure.ServiceStart) / time.Millisecond))
 			result.RequestsOriginated++
-		case <-workerCtx.Done():
+		case <-ctx.Done():
 			result.Duration = time.Since(start)
 			return result, nil
 		case <-statusTicker.C:
-			log.Printf("working... observations: %d", len(results))
+			log.Printf("working... current observation count: %d", len(results))
 		}
+
 	}
 }
 
-func produceMeasures(ctx context.Context, rate int, c chan<- benchMeasure) {
-	interval := time.Duration(float64(1) / float64(rate) * float64(time.Second))
-	ticker := time.NewTicker(interval)
-	defer ticker.Stop()
+type benchMeasure struct {
+	ResponseStart time.Time
+	ServiceStart  time.Time
+	End           time.Time
+	Error         error
+	ctx           context.Context
+	client        *client.TeleportClient
+	command       []string
+	interactive   bool
+}
+
+func run(ctx context.Context, request <-chan *benchMeasure, done chan<- *benchMeasure) {
+	defer func() {
+		if r := recover(); r != nil {
+			logrus.Warningf("recover from panic: %v", r)
+		}
+	}()
 	for {
 		select {
-		case <-ticker.C:
-
-			measure := benchMeasure{
-				Start: time.Now(),
-			}
-			select {
-			case c <- measure:
-			case <-ctx.Done():
+		case m, ok := <-request:
+			if !ok {
 				return
 			}
+			go work(ctx, m, done)
 		case <-ctx.Done():
 			return
 		}
 	}
 }
 
-type benchMeasure struct {
-	Start           time.Time
-	End             time.Time
-	ThreadCompleted bool
-	ThreadID        int
-	Error           error
-}
-
-type benchmarkThread struct {
-	id          int
-	ctx         context.Context
-	client      *client.TeleportClient
-	command     []string
-	interactive bool
-	receiveC    chan benchMeasure
-	sendC       chan benchMeasure
+func work(ctx context.Context, m *benchMeasure, send chan<- *benchMeasure) {
+	m.ServiceStart = time.Now()
+	execute(m)
```

Comment:
Perhaps consider having `execute` return an error (potentially nil) and then having `m` pass by value instead of by reference.  That would consolidate all of the modification to `m` into this this function, and let `execute` be more pure.